### ISSUE TITLE
getting absolute value of ciphertext

### ISSUE THREAD

**Issue by AndreiEn** [shaih/HElib#144]
Hello!
Let's say i have _a_ and _b_ encrypted mod some p (diffrent than 2, let's say 257). For a comparison between them, I saw in one of your comments that one should make _c=a-b_, and then to evaluate a mod p polynomial that return _1_ if _c>0_ and 0 otherwise. I don't really understand what did you mean by that, can you give more details ?. Or in other words, how can i evaluate if a ciphertext is greater than 0 (without encrypting the values bit by bit) ?. Or some idea to evaluate _|c|_ - the absolute value of a-b ?

**Comment by ssmiler** [comment]
You should look at Lagrangian interpolation. You can build a polynomial for
any function which gives the same value. In mod-p this polynomial will be
of degree p, thus big and costly to evaluate homomorphically.

**Comment by AndreiEn** [comment]
I looked into Lagrangian interpolation, but it won't work, because my number is either positive or negative. So let's say we get a value of -2 but we also can get 2, and for both cases it should return 2. And from what I understood from Lagrangian interpolation will give a wrong polynomial. Am i wrong ?

**Comment by ssmiler** [comment]
You should encode negative numbers into Z mod-p. For example 0..p/2-1 are
positive numbers from 0 to p/2-1 and p/2..p-1 are negative numbers
-p/2..-1. The interpolated polynomial should return the same value for
inputs between 0..p/2-1 and input minus p/2 otherwise.   Something like
this.

**Comment by AndreiEn** [comment]
You are right. Thx a lot. I still have one more question. This idea should work... but like you said it is too costly to evaluate it homomorphic (for more than 5-10 numbers). So what if i just decrypt the difference between the 2 numbers, and i use the value in clear for further computations. Would there be any major threat to security? Because I don't think there is any way to find the two values which produced that difference.

**Comment by ssmiler** [comment]
If one number is random it will be secure (it's like one-time pad ciphering). Other cases depend on how the two 
numbers are obtained.
